sefaz/accesskey/validation.go

```diff
@@ -0,0 +1,184 @@
+	if accessKey == "" {
+		return errors.E(op, ErrEmptyAccessKey, ErrCodeEmptyAccessKey)
+	}
+	if len(accessKey) != 44 {
+		return errors.E(op, ErrInvalidLenght, ErrCodeInvalidAccessKey)
+	}
+
+	if !isDigitOnly(accessKey) {
+		return errors.E(op, ErrInvalidCharacter, ErrCodeInvalidAccessKey)
+	}
+
+	if !isValidUF(accessKey[0:2]) {
+		return errors.E(op, ErrInvalidUF, ErrCodeInvalidAccessKey)
+	}
+
+	if !isValidMonth(accessKey[4:6]) {
+		return errors.E(op, ErrInvalidMonth, ErrCodeInvalidAccessKey)
+	}
+
+	if !isValidCPFCNPJ(accessKey[6:20].String()) {
+		return errors.E(op, ErrInvalidCPFCNPJ, ErrCodeInvalidAccessKey)
+	}
+
+	if !isValidModel(accessKey[20:22]) {
+		return errors.E(op, ErrInvalidModel, ErrCodeInvalidAccessKey)
+	}
+
+	if !isValidationDigitCorrect(accessKey) {
+		return errors.E(op, ErrInvalidDigit, ErrCodeInvalidAccessKey)
+	}
+
+	return nil
+}
+
+func (v *validator) Check(accessKey Accesskey) error {
+	return validate(accessKey)
+}
+
+func isDigitOnly(accesskey Accesskey) bool {
+	for _, token := range accesskey {
+		if !(token >= '0' && token <= '9') {
+			return false
+		}
+	}
+	return true
+}
+
+func isValidUF(uf Accesskey) bool {
+	if len(uf) != 2 {
+		return false
+	}
+	switch uf[0] {
+	case '1':
+		switch uf[1] {
+		case '0', '8', '9':
+			return false
+		default:
+			return true
+		}
+	case '2':
+		switch uf[1] {
+		case '0':
+			return false
+		default:
+			return true
+		}
+	case '3':
+		switch uf[1] {
+		case '1', '2', '3', '5':
+			return true
+		default:
+			return false
+		}
+	case '4':
+		switch uf[1] {
+		case '1', '2', '3':
+			return true
+		default:
+			return false
+		}
+	case '5':
+		switch uf[1] {
+		case '0', '1', '2', '3':
+			return true
+		default:
+			return false
+		}
+	default:
+		return false
+	}
+}
+
+func isValidMonth(month Accesskey) bool {
+	if len(month) != 2 {
+		return false
+	}
+	if month[0] == '0' {
+		return month[1] != '0'
+	}
+
+	if month[0] == '1' {
+		switch month[1] {
+		case '0', '1', '2':
+			return true
+		default:
+			return false
+		}
+	} else {
+		return false
+	}
+}
+
+func isValidModel(model Accesskey) bool {
+	return model == "55" || model == "57" ||
+		model == "65" || model == "67"
+}
+
+var validationDigitMultipliers = []int{
+	4, 3, 2, 9, 8, 7, 6, 5, 4, 3,
+	2, 9, 8, 7, 6, 5, 4, 3, 2, 9,
+	8, 7, 6, 5, 4, 3, 2, 9, 8, 7,
+	6, 5, 4, 3, 2, 9, 8, 7, 6, 5,
+	4, 3, 2,
+}
+
+func isValidationDigitCorrect(accessKey Accesskey) bool {
+	if len(accessKey) != 44 {
+		return false
+	}
+
+	var sum int
+
+	for i := 0; i < 43; i++ {
+		sum += (int(accessKey[i]) - '0') * validationDigitMultipliers[i]
+	}
+
+	r := sum % 11
+	vd := 0
+	if r > 1 {
+		vd = 11 - r
+	}
+
+	return (int(accessKey[43]) - '0') == vd
+}
+
+func isValidCPFCNPJ(cpfcnpj string) bool {
+	if stakeholder.CheckCNPJ(cpfcnpj) == nil {
+		return true
+	}
+
+	if cpfcnpj[0] != '0' || cpfcnpj[1] == '0' || cpfcnpj[2] == '0' {
+		return false
+	}
+
+	return stakeholder.CheckCPF(cpfcnpj[3:14]) == nil
+}
+
+/* isValidEmissionType
```

Comment:
Esse comentário está perdido aqui, se não há mais validação de tipo de emissão. Se for para deixar como histórico documentado, eu sugiro no tipo `AccessKey` ou em um readme.md embutido dentro desta pasta.

Reply:
Acho que o readme seria overkill, mas concordo que fica jogado do jeito que está. Vou mover para o arquivo accesskey.go